crates/wasi-provider/src/wasi_runtime.rs

```diff
@@ -71,64 +88,246 @@ impl WasiRuntime {
+                .unwrap()
+                .call(&[])
+            {
+                // We can't map errors here or it moves the send channel, so we
+                // do it in a match
+                Ok(m) => Ok(m),
+                Err(e) => {
+                    let message = "unable to run module";
+                    error!("{}: {:?}", message, e);
+                    status_sender
+                        .broadcast(ContainerState {
+                            terminated: Some(ContainerStateTerminated {
+                                message: Some(message.into()),
+                                reason: None,
+                                exit_code: 1,
+                                finished_at: Some(Time(chrono::Utc::now())),
+                                ..Default::default()
+                            }),
+                            ..Default::default()
+                        })
+                        .expect("status should be able to send");
+                    // Converting from anyhow
+                    Err(format_err!("{}: {}", message, e))
                 }
-            })
-            .collect::<Result<Vec<_>, _>>()?;
-
-        let instance = wasmtime::Instance::new(&module, &imports)
-            .map_err(|e| format_err!("unable to instantiate module: {}", e))?;
-
-        // NOTE(taylor): In the future, if we want to pass args directly, we'll
-        // need to do a bit more to pass them in here.
-        info!("starting run of module");
-        instance
-            .get_export("_start")
-            .expect("export")
-            .func()
-            .unwrap()
-            .call(&[])
-            .unwrap();
-
-        info!("module run complete");
-        Ok(())
+            }?;
+
+            info!("module run complete");
+            status_sender
+                .broadcast(ContainerState {
+                    terminated: Some(ContainerStateTerminated {
+                        message: Some("Module run completed".into()),
+                        exit_code: 0,
+                        finished_at: Some(Time(chrono::Utc::now())),
+                        ..Default::default()
+                    }),
+                    ..Default::default()
+                })
+                .expect("status should be able to send");
+            Ok(())
+        });
+
+        Ok(RuntimeHandle::new(
+            tokio::fs::File::from_std(self.output.reopen()?),
+            handle,
+            status_recv,
+        ))
+    }
+}
+
+/// Represents a handle to a running WASI instance. Right now, this is
+/// experimental and just for use with the [crate::WasiProvider]. If we like
+/// this pattern, we will expose it as part of the kubelet crate
+pub struct RuntimeHandle<R: AsyncReadExt + AsyncSeekExt + Unpin> {
+    output: BufReader<R>,
+    handle: JoinHandle<Result<(), failure::Error>>,
+    status_channel: Receiver<ContainerState>,
+}
+
+impl<R: AsyncReadExt + AsyncSeekExt + Unpin> RuntimeHandle<R> {
```

Comment:
I believe we can just require `AsyncRead` and `AsyncSeek` since there are blanket impls for the extension traits.

Reply:
Without using the `Ext` traits, I couldn't access the `read_to_end` or `seek` methods